src/library/export/engineprimeexportjob.cpp

```diff
@@ -0,0 +1,534 @@
+        djinterop::database& db,
+        QHash<TrackId, int64_t>& mixxxToEnginePrimeTrackIdMap,
+        const TrackPointer pTrack,
+        std::unique_ptr<Waveform> pWaveform) {
+    // Only export supported file types.
+    if (!kSupportedFileTypes.contains(pTrack->getType())) {
+        qInfo() << "Skipping file" << pTrack->getFileInfo().fileName()
+                << "(id" << pTrack->getId() << ") as its file type"
+                << pTrack->getType() << "is not supported";
+        return;
+    }
+
+    // Copy the file, if required.
+    auto musicFileRelativePath = exportFile(request, pTrack);
+
+    // Export meta-data.
+    exportMetadata(db, mixxxToEnginePrimeTrackIdMap, pTrack, std::move(pWaveform), musicFileRelativePath);
+}
+
+void exportCrate(
+        djinterop::crate& extRootCrate,
+        QHash<TrackId, int64_t>& mixxxToEnginePrimeTrackIdMap,
+        const Crate& crate,
+        const QList<TrackId>& trackIds) {
+    // Create a new crate as a sub-crate of the top-level Mixxx crate.
+    auto extCrate = extRootCrate.create_sub_crate(crate.getName().toStdString());
+
+    // Loop through all track ids in this crate and add.
+    for (auto iter = trackIds.cbegin(); iter != trackIds.cend(); ++iter) {
+        auto trackId = *iter;
+        auto extTrackId = mixxxToEnginePrimeTrackIdMap[trackId];
+        extCrate.add_track(extTrackId);
+    }
+}
+
+} // namespace
+
+EnginePrimeExportJob::EnginePrimeExportJob(
+        QObject* parent,
+        TrackCollectionManager* pTrackCollectionManager,
+        EnginePrimeExportRequest request)
+        : QThread(parent),
+          m_pTrackCollectionManager(pTrackCollectionManager),
+          m_request{std::move(request)} {
+    // Must be collocated with the TrackCollectionManager.
+    DEBUG_ASSERT(m_pTrackCollectionManager != nullptr);
+    moveToThread(m_pTrackCollectionManager->thread());
+}
+
+void EnginePrimeExportJob::loadIds(QSet<CrateId> crateIds) {
+    // Note: this slot exists to ensure the track collection is never accessed
+    // from outside its own thread.
+    DEBUG_ASSERT(thread() == m_pTrackCollectionManager->thread());
+    QMutexLocker lock{&m_mainThreadLoadMutex};
+
+    if (crateIds.isEmpty()) {
+        // No explicit crate ids specified, meaning we want to export the
+        // whole library, plus all non-empty crates.  Start by building a list
+        // of unique track refs from all directories in the library.
+        qDebug() << "Loading all track refs and crate ids...";
+        QSet<TrackRef> trackRefs;
+        auto dirs = m_pTrackCollectionManager->internalCollection()
+                            ->getDirectoryDAO()
+                            .getDirs();
+        for (auto& dir : dirs) {
+            trackRefs.unite(m_pTrackCollectionManager->internalCollection()
+                                    ->getTrackDAO()
+                                    .getAllTrackRefs(dir)
+                                    .toSet());
```

Comment:
```
/home/jan/Projects/mixxx/src/library/export/engineprimeexportjob.cpp: In member function ‘void mixxx::EnginePrimeExportJob::loadIds(QSet<CrateId>)’:
/home/jan/Projects/mixxx/src/library/export/engineprimeexportjob.cpp:319:44: warning: ‘QSet<T> QList<T>::toSet() const [with T = TrackRef]’ is deprecated: Use QSet<T>(list.begin(), list.end()) instead. [-Wdeprecated-declarations]
  319 |                                     .toSet());
      |                                            ^
In file included from /usr/include/qt/QtCore/QSet:1,
                 from /home/jan/Projects/mixxx/src/library/export/engineprimeexportjob.h:7,
                 from /home/jan/Projects/mixxx/src/library/export/engineprimeexportjob.cpp:1:
/usr/include/qt/QtCore/qset.h:402:30: note: declared here
  402 | Q_OUTOFLINE_TEMPLATE QSet<T> QList<T>::toSet() const
      |                              ^~~~~~~~
```

Reply:
@Holzhaus what is the minimum version of QT required to build Mixxx?  The [docs](https://doc.qt.io/qt-5/qlist.html#toSet) for `QList<T>::toSet()` say:

> Note: Since Qt 5.14, range constructors are available for Qt's generic container classes and should be used in place of this method.

(I've am building with QT 5.12.8 on my development machine, which is presumably why I haven't seen this warning before...)

Reply:
Qt 5.12 for the moment. Have a look at https://github.com/mixxxdj/mixxx/commit/5f47c3d4c444087ef56396e9319721cdeadd3afc for a workaround.

Reply:
Ah, I thought I might have to do something like that :smile:.  Will push an update shortly.

Reply:
@Holzhaus  I've pushed an update along the lines of the workaround mentioned (and rebased onto latest master).  Tested on Linux Mint 20 (with QT 5.12.8) and Fedora 32 (with QT 5.14.2).